### PR TITLE
 FEATURE: new date/time components

### PR DESCRIPTION
**Pull request by jjaffeux**
Usage:

```
{{time-input hours=hours minutes=minutes onChange=onChange}}
```

```
{{date-input date=date onChange=onChange}}
```

```
{{date-time-input date=date onChange=onChange showTime=ShowTime}}
```

```
{{date-time-input-range
  from=fromDate
  to=toDate
  showFromTime=showFromTime
  showToTime=showToTime
  onChange=onChange}}
```

Screenshot 2019-07-17 at 09 11 34

There's also a discourse-styleguide PR coming with it: https://github.com/discourse/discourse-styleguide/pull/17

### CODE REVIEW

File: app/assets/javascripts/discourse/components/time-input.js.es6 [hunk 1]
```diff
@@ -0,0 +1,63 @@
+import { isNumeric } from "discourse/lib/utilities";
+// const TIME_REGEX = /(\d{1,2}):(\d{1,2})/;
```
Comment (ZogStriP) [inline app/assets/javascripts/discourse/components/time-input.js.es6]: Might want to remove that ;)

File: app/assets/javascripts/discourse/controllers/application.js.es6 [hunk 2]
```diff
@@ -22,5 +22,39 @@ export default Ember.Controller.extend({
...
+  fromHours: null,
+  toHours: null,
+
+  init() {
+    this.setProperties({ from: new Date(), to: new Date() });
+  },
+
+  actions: {
+    onChangeFromDate(date) {
+      console.log("onChangeFromDate", date);
```
Comment (ZogStriP) [inline app/assets/javascripts/discourse/controllers/application.js.es6]: Those should not be needed anymore ;)

File: app/assets/javascripts/discourse/templates/components/time-input.hbs [hunk 3]
```diff
@@ -0,0 +1,35 @@
+<div class="fields">
+  {{#if nativePicker}}
+    {{!-- {{input class="field time" type="time" value=_time}} --}}
```
Comment (ZogStriP) [inline app/assets/javascripts/discourse/templates/components/time-input.hbs]: Should this be commented out?

File: app/assets/javascripts/discourse/templates/components/time-input.hbs [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+      maxlength=2
+      max="23"
+      min="0"
+      placeholder="00"
+      value=_hours
+      input=(action "onInput" (hash prop="hours"))
+      focus-in=(action "onFocusIn")
+    }}
+
+    <div class="separator">:</div>
```
Comment (ZogStriP) [inline app/assets/javascripts/discourse/templates/components/time-input.hbs]: I wonder, is time universally written as `HH:mm` or are there locales where it isn't?

**Comment by jjaffeux** [comment]
opps I thought I put draft sorry. It’s almost ready yes, thx!

File: app/assets/javascripts/discourse/templates/components/time-input.hbs (hunk 4 above)
Comment (jjaffeux) [reply app/assets/javascripts/discourse/templates/components/time-input.hbs]: I would want to start with this, and improve along the way as I implement it in the various UIs/cases we have using it

**Comment by jjaffeux** [comment]
@ZogStriP should be good

File: app/assets/stylesheets/common/components/date-input.scss [hunk 5]
```diff
@@ -0,0 +1,13 @@
+.d-date-input {
+  .date-picker {
+    margin: 0;
+    text-align: left;
+    width: 100%;
+    outline: none;
+    box-shadow: none !important;
```
Comment (eviltrout) [inline app/assets/stylesheets/common/components/date-input.scss]: :cry:

File: config/locales/client.en.yml [hunk 6]
```diff
@@ -1518,6 +1518,12 @@ en:
         one: "Select at least {{count}} item."
         other: "Select at least {{count}} items."
 
+    date_time_picker:
+      from: From
+      to: To
+      errors:
+        to_before_from: "To date must be set after from date."
```
Comment (eviltrout) [inline config/locales/client.en.yml]:
```suggestion
        to_before_from: "To date must be later than from date."
```

File: app/assets/stylesheets/common/components/date-input.scss (hunk 5 above)
Comment (jjaffeux) [reply app/assets/stylesheets/common/components/date-input.scss]: yes agreed, my feeling is that I will be able to remove old pickers in the future and then remove this, so Im more eager to accept it 🤷‍♂️

**Pull request merged**